tools/urtrace/urtrace

```diff
@@ -0,0 +1,136 @@
+parser.add_argument("--begin", help="Print on function begin.", action="store_true")
+parser.add_argument("--time-unit", choices=['ns', 'us', 'ms', 's', 'auto'], default='auto', help="Use a specific unit of time for profiling.")
+parser.add_argument("--libpath", default=['.', '../lib/', '/lib/', '/usr/local/lib/', '/usr/lib/'], action="append", help="Search path for adapters and xpti libraries.")
+parser.add_argument("--recursive", help="Use recursive library search.", action="store_true")
+parser.add_argument("--debug", help="Print tool debug information.", action="store_true")
+args = parser.parse_args()
+config = vars(args)
+if args.debug:
+    print(config)
+env = dict()
+
+env['UR_LOG_COLLECTOR']="level:info;output:stderr;flush:info"
+collector_args = ""
+if args.begin:
+    collector_args += "print_begin;"
+if args.profiling:
+    collector_args += "profiling;"
+if args.time_unit:
+    collector_args += "time_unit:" + args.time_unit + ";"
+if args.filter:
+    collector_args += "filter:" + args.filter + ";"
+if args.no_args:
+    collector_args += "no_args;"
+env['UR_COLLECTOR_ARGS'] = collector_args
+
+log_collector = ""
+if args.debug:
+    log_collector += "level:debug;"
+else:
+    log_collector += "level:info;"
+if args.file:
+    log_collector += "output:file," + args.file + ";"
+elif args.stdout:
+    log_collector += "output:stdout;"
+else:
+    log_collector += "output:stderr;"
+log_collector += "flush:error"
+env['UR_LOG_COLLECTOR'] = log_collector
+
+env['XPTI_TRACE_ENABLE'] = "1"
+
+xptifw_lib = get_dynamic_library_name("xptifw")
+
+xptifw = find_library(args.libpath, xptifw_lib, args.recursive)
+if xptifw is None:
+    sys.exit("unable to find xptifw library - " + xptifw_lib)
+env['XPTI_FRAMEWORK_DISPATCHER'] = xptifw
+
+collector_lib = get_dynamic_library_name("ur_collector")
+
+collector = find_library(args.libpath, collector_lib, args.recursive)
+if collector is None:
+    sys.exit("unable to find collector library - " + collector_lib)
+env['XPTI_SUBSCRIBERS'] = collector
+
+force_load = None
+
+if args.null:
+    null_lib = get_dynamic_library_name("ur_adapter_null")
+    null_adapter = find_library(args.libpath, null_lib, args.recursive)
+    if null_adapter is None:
+        sys.exit("unable to find the null adapter - " + null_lib)
+    force_load = null_adapter
+
+for adapter in args.adapter:
+    adapter_path = find_library(args.libpath, adapter, args.recursive) if is_filename(adapter) else adapter
+    if adapter_path is None:
+        sys.exit("adapter does not specify a valid file " + args.adapter)
+    if force_load is not None:
+        force_load += "," + adapter_path
+    else:
+        force_load = adapter_path
+
+if force_load:
+    env['UR_ADAPTERS_FORCE_LOAD'] = force_load
+
+if args.debug:
+    print(env)
+
+result = subprocess.run(config['command'], env=env)
+if args.debug:
+    print(result)
+
+exit(result.returncode)
```

Comment:
No empty line at EOF.

Reply:
That's not true? Git shows something like 'no newline at end of file' and on github in PRs there's a red bar if that's the case.

Reply:
That's what I meant, to add an empty newline, sorry for misconfusion.